test/acvp_kyber.c

```diff
@@ -0,0 +1,269 @@
+
+typedef enum { AFT, VAL } acvp_type;
+
+typedef enum { encapsulation, decapsulation } acvp_encapDecap_function;
+
+/* Decode hex character [0-9A-Fa-f] into 0-15 */
+static unsigned char decode_hex_char(char hex) {
+  if (hex >= '0' && hex <= '9') {
+    return (unsigned char)(hex - '0');
+  } else if (hex >= 'A' && hex <= 'F') {
+    return 10 + (unsigned char)(hex - 'A');
+  } else if (hex >= 'a' && hex <= 'f') {
+    return 10 + (unsigned char)(hex - 'a');
+  } else {
+    return 0xFF;
+  }
+}
+
+static int decode_hex(const char *prefix, unsigned char *out, size_t out_len,
+                      const char *hex) {
+  size_t hex_len = strlen(hex);
+  size_t prefix_len = strlen(prefix);
+
+  // Check that hex starts with `prefix=`
+  // Use memcmp, not strcmp
+  if (hex_len < prefix_len + 1 || memcmp(prefix, hex, prefix_len) != 0 ||
+      hex[prefix_len] != '=') {
+    goto hex_usage;
+  }
+
+  hex += prefix_len + 1;
+  hex_len -= prefix_len + 1;
+
+  if (hex_len != 2 * out_len) {
+    goto hex_usage;
+  }
+
+  for (size_t i = 0; i < out_len; i++, hex += 2, out++) {
+    unsigned hex0 = decode_hex_char(hex[0]);
+    unsigned hex1 = decode_hex_char(hex[1]);
+    if (hex0 == 0xFF || hex1 == 0xFF) {
+      goto hex_usage;
+    }
+
+    *out = (hex0 << 4) | hex1;
+  }
+
+  return 0;
+
+hex_usage:
+  fprintf(stderr,
+          "Argument %s invalid: Expected argument of the form '%s=HEX' with "
+          "HEX being a hex encoding of %u bytes\n",
+          hex, prefix, (unsigned)out_len);
+  return 1;
+}
+
+static void print_hex(const char *name, const unsigned char *raw, size_t len) {
+  if (name != NULL) {
+    printf("%s=", name);
+  }
+  for (; len > 0; len--, raw++) {
+    printf("%02X", *raw);
+  }
+  printf("\n");
+}
+
+static void acvp_kyber_encapDecp_AFT_encapsulation(
+    unsigned char const ek[KYBER_INDCPA_PUBLICKEYBYTES],
+    unsigned char const m[KYBER_SYMBYTES]) {
+  unsigned char ct[KYBER_CIPHERTEXTBYTES];
+  unsigned char ss[KYBER_SSBYTES];
+
+  crypto_kem_enc_derand(ct, ss, ek, m);
+
+  print_hex("ct", ct, sizeof(ct));
+  print_hex("ss", ss, sizeof(ss));
+}
+
+static void acvp_kyber_encapDecp_VAL_decapsulation(
+    unsigned char const dk[KYBER_SECRETKEYBYTES],
+    unsigned char const c[KYBER_CIPHERTEXTBYTES]) {
+  unsigned char ss[KYBER_SSBYTES];
+
+  crypto_kem_dec(ss, c, dk);
+
+  print_hex("ss", ss, sizeof(ss));
+}
+
+static void acvp_kyber_keyGen_AFT(unsigned char const z[KYBER_SYMBYTES],
+                                  unsigned char const d[KYBER_SYMBYTES]) {
+  unsigned char ek[KYBER_INDCPA_PUBLICKEYBYTES];
+  unsigned char dk[KYBER_SECRETKEYBYTES];
+
+  unsigned char zd[2 * KYBER_SYMBYTES];
+  memcpy(zd, d, KYBER_SYMBYTES);
+  memcpy(zd + KYBER_SYMBYTES, z, KYBER_SYMBYTES);
+
+  crypto_kem_keypair_derand(ek, dk, zd);
+
+  print_hex("ek", ek, sizeof(ek));
+  print_hex("dk", dk, sizeof(dk));
+}
+
+int main(int argc, char *argv[]) {
+  if (argc == 0) {
+    goto usage;
+  }
+  argc--, argv++;
```

Comment:
I find this very confusing.  can we just keep argc and argv constant?

Reply:
Again, why?

Reply:
No action.